packages/flutter_tools/lib/src/base/deferred_component.dart

```diff
@@ -0,0 +1,216 @@
+  /// packaged in this component. However, libraries that are not listed here
+  /// may also be included if the loading units that are needed also contain
+  /// libraries that are not listed here.
+  final List<String> libraries;
+
+  /// Assets that are part of this component as a Uri relative to the project directory.
+  final List<Uri> assets;
+
+  /// The minimal set of [LoadingUnit]s needed that contain all of the dart libraries in
+  /// [libraries].
+  ///
+  /// Each [LoadingUnit] contains the compiled code for a set of dart libraries. Each
+  /// [DeferredComponent] contains a list of dart libraries that must be included in the
+  /// component. The set [loadingUnits] is all of the [LoadingUnit]s needed such that
+  /// all required dart libs in [libraries] are in the union of the [LoadingUnit.libraries]
+  /// included by the loading units in [loadingUnits].
+  ///
+  /// When [loadingUnits] is non-null, then the component is considered [assigned] and the
+  /// field [assigned] will be true. When [loadingUnits] is null, then the component is
+  /// unassigned and should not be used for any tasks that require loading unit information.
+  /// When using [loadingUnits], [assigned] should be checked first. Loading units can be
+  /// assigned with [assignLoadingUnits].
+  Set<LoadingUnit> get loadingUnits => _loadingUnits;
+  Set<LoadingUnit> _loadingUnits;
+
+  /// Indicates if the component has loading units assigned.
+  ///
+  /// Unassigned components simply reflect the pubspec.yaml configuration directly,
+  /// contain no loading unit data, and [loadingUnits] is null. Once assigned, the component
+  /// will contain a set of [loadingUnits] which contains the [LoadingUnit]s that the
+  /// component needs to include. Loading units can be assigned with the [assignLoadingUnits]
+  /// call.
+  bool get assigned => _assigned;
+  bool _assigned;
+
+  /// Selects the [LoadingUnit]s that contain this component's dart libraries.
+  ///
+  /// After calling this method, this [DeferredComponent] will be considered [assigned],
+  /// and [loadingUnits] will return a non-null result.
+  ///
+  /// [LoadingUnit]s in `allLoadingUnits` that contain libraries that are in [libraries]
+  /// are added to the set [loadingUnits].
+  ///
+  /// Providing null or empty list of `allLoadingUnits` will still change the assigned
+  /// status, but will result in [loadingUnits] returning an empty set.
+  void assignLoadingUnits(List<LoadingUnit> allLoadingUnits) {
+    _assigned = true;
+    _loadingUnits = <LoadingUnit>{};
+    if (allLoadingUnits == null) {
+      return;
+    }
+    for (final String lib in libraries) {
+      for (final LoadingUnit loadingUnit in allLoadingUnits) {
+        if (loadingUnit.libraries.contains(lib)) {
+          _loadingUnits.add(loadingUnit);
+        }
+      }
+    }
+  }
+
+  /// Provides a human readable string representation of the
+  /// configuration.
+  @override
+  String toString() {
```

Comment:
This method is non-trivial but also has no unit test coverage. What is this intended for, just debugging?

Reply:
In the setup validator (coming in a future PR, currently in https://github.com/flutter/flutter/pull/63773), these toStrings are used to display to the developer what changes/loading units were generated when validation fails.

I'll add tests for these.